pgml-cms/blog/our-migration-from-aws-to-gcp-with-minimal-downtime.md

```diff
@@ -0,0 +1,134 @@
+---
+description: >-
+  Lessons learned from moving terabytes of real time data between cloud providers.
+featured: false
+tags: [engineering]
+---
+
+# Our migration from AWS to GCP with minimal downtime
+
+<div align="left">
+
+<figure><img src=".gitbook/assets/lev.jpg" alt="Author" width="125"><figcaption></figcaption></figure>
+
+</div>
+
+Lev Kokotov
+
+June 6, 2024
+
+From the beginning, our plan for PostgresML was to be cloud-agnostic. Since we are an infrastructure provider, we have to deploy our code where our customers are. Like most startups, we started on AWS, because that's what we knew best. After over 10 years of AWS experience, and its general dominance in the market, it seemed right to build something we've done before, this time in Rust of course.
+
+After talking to several customers, we've noticed a pattern: most of them were using either Azure or GCP. So we had to go back to our original plan. Our platform manages all infrastructure internally, by representing common concepts like hosts, networking rules, open ports, and domains as first class entities in our codebase. To add additional cloud vendors, we just had to write integrations with their APIs.
+
+## Cloud-agnostic from the start
+
+PostgresML, much like Postgres itself, can run on a variety of platforms. Our operating system of choice, **Ubuntu**, is available on all clouds, and comes with good support for GPUs. We therefore had no trouble spinning up machines on Azure and GCP with identical software to match our AWS deployments.
+
+Since we're first and foremost a database company, data integrity and security are extremely important. To achieve that goal, and to be independent from any cloud-specific storage solutions, we are using **ZFS** as our filesystem to store Postgres data.
+
+Moving ZFS filesystems between machines is a solved problem, or so we thought.
+
+## The migration
+
+Our primary Serverless deployment was in Oregon, AWS *us-west-2* region. We were moving it to GCP in Iowa, *us-central1* region.
```

Comment:
`-` Our primary Serverless deployment was in Oregon, AWS *us-west-2* region. We were moving it to GCP in Iowa, *us-central1* region.


`+` Much of our Serverless infrastructure was deployed in Oregon, AWS *us-west-2* region. We were moving it to GCP in Iowa, *us-central1* region to be more centrally located with respect to US datacenters.